platforms/android/run_module_app/src/main.cc

```diff
@@ -0,0 +1,209 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <android/log.h>
+#include <android_native_app_glue.h>
+
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+#include "iree/base/initializer.h"
+#include "iree/base/status.h"
+#include "iree/modules/hal/hal_module.h"
+#include "iree/tools/vm_util.h"
+#include "iree/vm/api.h"
+
+namespace iree {
+namespace {
+
+const char* kAppTag = "iree-run-module";
+#define LOGI(...) \
+  ((void)__android_log_print(ANDROID_LOG_INFO, kAppTag, __VA_ARGS__))
+#define LOGE(...) \
+  ((void)__android_log_print(ANDROID_LOG_ERROR, kAppTag, __VA_ARGS__))
+
+const char kModuleFileName[] = "module.vmfb";
+const char kEntryFunctionFileName[] = "entry_function.txt";
+const char kInputsFileName[] = "inputs.txt";
+const char kDriverFileName[] = "driver.txt";
+
+// A struct containing information regarding one IREE VM module invocation.
+struct IreeModuleInvocation {
+  std::string module;
+  std::string entry_function;
+  std::string inputs;
+  std::string driver;
+
+  operator bool() const {
+    return !module.empty() && !entry_function.empty() && !inputs.empty() &&
+           !driver.empty();
+  }
+};
+
+// A class for loading IREE module invocation information from Android apk asset
+// files.
+class ModuleLoader {
+ public:
+  explicit ModuleLoader(android_app* app) : app_context_(app) {}
+  ~ModuleLoader() = default;
+
+  IreeModuleInvocation LoadModuleInvocation() {
+    IreeModuleInvocation invocation = {};
+    invocation.module = ReadFileAsset(kModuleFileName);
+    invocation.entry_function = ReadFileAsset(kEntryFunctionFileName);
+    invocation.inputs = ReadFileAsset(kInputsFileName);
+    invocation.driver = ReadFileAsset(kDriverFileName);
+    return invocation;
+  }
+
+ private:
+  // Reads the given asset file and returns its contents. Returns an empty
+  // string on error.
```

Comment:
Use `StatusOr<std::string>` instead of returning an empty string?

Reply:
It would be nice to have llvm::Optional here. StatusOr is a bit overkill to me given we don't have more than one error state, which an empty string can indicate.

Reply:
> It would be nice to have llvm::Optional here.

We have access to absl::optional: https://github.com/abseil/abseil-cpp/blob/master/absl/types/optional.h. See also https://abseil.io/tips/123.

> StatusOr is a bit overkill to me given we don't have more than one error state, which an empty string can indicate.

Empty string doesn't communicate or propagate errors as clearly as `StatusOr<std::string>` does. This code logs errors then continues on logging empty strings and only later fails when `LoadBytecodeModule` or one of the other later functions is called. Status lets you more expressively track and handle failure states. [1]

This matters less, but there could later be multiple error states (permission denied, file not found, resource busy, network disconnected, etc.). Using Status now lets you keep a fixed API that supports multiple types of errors from the start, potentially saving some refactoring work later.

Please switch to `StatusOr` and also update the call signature on `LoadModuleInvocation` to propagate errors. Also, the rest of this file is already using Status - we shouldn't mix error handling styles if we can avoid it.

---

[1] One case that Status is awkward for that could be relevant here: Status typically fails fast, but you might want to try reading all files, then aggregating their errors and failing if any of them could not be read.

So rather than have
```c++
    IREE_ASSIGN_OR_RETURN(invocation.module, ReadFileAsset(kModuleFileName));
    IREE_ASSIGN_OR_RETURN(invocation.entry_function,
                          ReadFileAsset(kEntryFunctionFileName));
    IREE_ASSIGN_OR_RETURN(invocation.inputs, ReadFileAsset(kInputsFileName));
    IREE_ASSIGN_OR_RETURN(invocation.driver, ReadFileAsset(kDriverFileName));
```

You could imagine something like this:
```c++
    Status last_status;
    if (auto module = ReadFileAsset(kModuleFileName); module.ok()) {
      invocation.module = module.value();
    } else {
      last_status = module.status();
    }
    ...
    return last_status;
```

(the code here sort of does this through its logging, but at the cost of complete error handling)

Reply:
Makes sense to me! Changed. :)